package wishful_controller.framework;

message CmdDesc {
    required string type = 1;
    optional string family = 2;
    required string func_name = 3;
    optional string exec_time = 4;
    optional string call_id = 5;
    optional string caller_id = 6;
    optional string transaction_id = 7;
    optional string interface = 8;

    enum Serialization {
      NONE = 0;
      PICKLE = 1;
      PROTOBUF = 2;
      JSON = 3;
    }
    optional Serialization serialization_type = 9 [default = NONE];

    optional string repeat_interval = 10;
    optional uint32 repeat_number = 11;
}

message Device {
    required uint32 id = 1;
    required string name = 2;
}

message Module {
    message Attribute {
        required string name = 1;
    }
    message Function {
        required string name = 1;
    }
    message Event {
        required string name = 1;
    }
    message Service {
        required string name = 1;
    }
    required uint32 id = 1;
    required string name = 2;
    optional Device device = 3;
    repeated Attribute attributes = 4;
    repeated Function functions = 5;
    repeated Event events = 6;
    repeated Service services = 7;
}

message NodeInfoMsg {
    required string agent_uuid = 1;
    required string ip = 2;
    optional string name = 3;
    optional string info = 4;
    repeated Module modules = 5;
}

message NodeInfoRequest {
    required string agent_uuid = 1;
}

message NodeAddNotification {
    required string agent_uuid = 1;
}

message NewNodeAck {
    required bool status = 1;
    optional string controller_uuid = 2;
    optional string agent_uuid = 3;
    repeated string topics = 4;
}

message NodeExitMsg {
    required string agent_uuid = 1;
    optional string reason = 2;
}

message HelloMsg {
    required string uuid = 1;
    required uint32 timeout = 2;
}
